### PR TITLE
Add resume service support to supervisor [ECR-3773]

### PR DESCRIPTION
**Pull request by alekseysidorov**

### CODE REVIEW

File: services/supervisor/src/transactions.rs [hunk 1]
```diff
@@ -131,6 +128,47 @@ impl StopService {
     }
 }
 
+impl ResumeService {
+    fn validate(&self, context: &CallContext<'_>) -> Result<(), ExecutionError> {
+        let instance = get_instance(context, self.instance_id)?;
+
+        match instance.status {
```
Comment (popzxc) [inline services/supervisor/src/transactions.rs]: Nit: you can just write `if instance.status != InstanceStatue::Stopped`

Comment (aleksuss) [reply services/supervisor/src/transactions.rs]: Really ? `Option` derives `PartialOrd` and `Eq`. Take a look [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b3ff083e7147f89c341020d7392a1f31)

File: services/supervisor/src/transactions.rs [hunk 2]
```diff
@@ -131,6 +128,47 @@ impl StopService {
...
+                ConfigurationError::MalformedConfigPropose.with_description(format!(
+                    "Discarded an attempt to resume service with incorrect artifact version: \
+                     got {}, expected {}",
+                    self.artifact.version,
+                    instance.data_version()
+                )),
+            );
+        }
+
+        if instance.spec.artifact.name != self.artifact.name {
```
Comment (slowli) [inline services/supervisor/src/transactions.rs]: Nit: It could make sense to use [`ArtifactId::is_upgrade_of`](https://github.com/exonum/exonum/blob/616622f3b7e7bf7858590bae6b11732704fcfd7c/exonum/src/runtime/types.rs#L201) to be more forward-compatible. OTOH, the error message will be less specific in this case.

File: services/supervisor/src/proto_structures.rs [hunk 3]
```diff
@@ -124,6 +137,8 @@ pub enum ConfigChange {
     StartService(StartService),
     /// Existing service instance stop request.
     StopService(StopService),
+    /// Previously stopped service instance resume request.
```
Comment (slowli) [inline services/supervisor/src/proto_structures.rs]: Nit: IMO, "... request" descriptions are weirdly phrased. I'd suggest something like "Request to resume a previously stopped service instance."

File: services/supervisor/src/proto/service.proto [hunk 4]
```diff
@@ -65,6 +65,16 @@ message ServiceConfig {
   bytes params = 2;
 }
 
+// Request for the resume previously stopped service instance.
```
Comment (slowli) [inline services/supervisor/src/proto/service.proto]:
```suggestion
// Request to resume a previously stopped service instance.
```

File: components/crypto/src/lib.rs [hunk 5]
```diff
@@ -469,7 +469,7 @@ pub struct KeyPair {
 
 impl KeyPair {
     pub fn from_keys(public_key: PublicKey, secret_key: SecretKey) -> Self {
-        debug_assert!(
+        assert!(
```
Comment (slowli) [inline components/crypto/src/lib.rs]: Why was this changed?

File: services/supervisor/src/transactions.rs (hunk 1 above)
Comment (alekseysidorov) [reply services/supervisor/src/transactions.rs]: You are right, today I am a little bit stupid.

File: CHANGELOG.md [hunk 6]
```diff
@@ -118,6 +118,10 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
 - Error code values were changed and split into several enum
   representing sub-groups. (#1680)
 
+- Added support for resuming previously stopped services. (#1706)
```
Comment (slowli) [inline CHANGELOG.md]: This isn't a breaking change, is it?

File: services/supervisor/tests/supervisor/service_lifecycle.rs [hunk 7]
```diff
@@ -178,3 +177,128 @@ fn stop_already_stopped_service() {
             )
     )
 }
+
+#[test]
+fn resume_stopped_service() {
```
Comment (slowli) [inline services/supervisor/tests/supervisor/service_lifecycle.rs]: Nit: Could you add a test to check that it's impossible to request resume and some other operation regarding the service (e.g., changing its config) in the same update?

File: components/crypto/src/lib.rs (hunk 5 above)
Comment (alekseysidorov) [reply components/crypto/src/lib.rs]: Because this code breaks tests in release mode and @aleksuss asked me to fix this micro issue.

File: services/supervisor/src/transactions.rs (hunk 2 above)
Comment (alekseysidorov) [reply services/supervisor/src/transactions.rs]: `is_upgrade_of` has a little bit different behaviour.
